README.rst
===========
kotti_media
===========

This is an extension to the Kotti CMS that allows you to add a quiz content type to your Kotti site.

`Find out more about Kotti`_

``kotti_quiz`` lets you create quizzes with freetext, single-choice and multi-choice answers.

Setup
=====

To activate the ``kotti_quiz`` add-on in your Kotti site, you need to add an entry to the ``kotti.configurators`` setting in your Paste Deploy config.
If you don't have a ``kotti.configurators`` option, add one.
The line in your ``[app:main]`` section could then look like this::

  kotti.configurators = kotti_quiz.kotti_configure

With this, you'll be able to add quiz items in your site.

In your settings file, set kotti_quiz.asset_overrides to a list of asset specifications. This allows you to set up a directory in your package that will mirror kotti_quiz’s own and that allows you to override kotti_quiz’s templates on a case by case basis.

Usage
=====

A standard way to use kotti_media is to first create a Document in your content tree; this document will become the quiz section of your application. Then you can add childs to your quiz section by appending quiz content to it.


Work in progress
================

``kotti_quiz`` is considered alpha software, not yet suitable for use in production environments.
The current state of the project is in no way feature complete nor API stable.
If you really want to use it in your project(s), make sure to pin the exact version in your requirements.
Not doing so will likely break your project when future releases become available.


Development
===========

Contributions to ``kotti_quiz`` are highly welcome.
Just clone its `Github repository`_ and submit your contributions as pull requests.

Testing
-------

|build status|_

``kotti_quiz`` has 100% test coverage.
Please make sure that you add tests for new features and that all tests pass before submitting pull requests.
Running the test suite is as easy as running ``py.test`` from the source directory (you might need to run ``python setup.py dev`` to have all the test requirements installed in your virtualenv).


.. _Find out more about Kotti: http://pypi.python.org/pypi/Kotti
.. _Github repository: https://github.com/sbabrass/kotti_quiz
.. _gitflow: https://github.com/nvie/gitflow
.. _A successful Git branching model: http://nvie.com/posts/a-successful-git-branching-model/
.. _MediaElementJS: http://mediaelementjs.com/
.. |build status| image:: https://secure.travis-ci.org/sbabrass/kotti_quiz.png?branch=master
.. _build status: http://travis-ci.org/sbabrass/kotti_quiz


CHANGES.rst
Changelog
=========

0.1-dev (unreleased)
--------------------

- Package created using templer
  [Andreas Kaiser]

kotti_quiz/tests/test_functional_content_type.py
from kotti.resources import get_root
from kotti.testing import DummyRequest
from kotti.testing import FunctionalTestBase
from kotti.testing import setUpFunctional

from kotti_quiz.resources import ContentType
from kotti_quiz.views import view_content_type


class TestContentView(FunctionalTestBase):

    def setUp(self, **kwargs):
        settings = {'kotti.configurators': 'kotti_quiz.kotti_configure'}
        self.__dict__.update(setUpFunctional(**settings))

    def test_content_view(self):
        root = get_root()
        root['content-type'] = ContentType(title=u'New Content Type',
                                           example_text=u'some text')
        result = view_content_type(root['content-type'], DummyRequest())
        assert result[u'example_text'] == u'some text'

    def test_browser_call(self):
        browser = self.login_testbrowser()
        browser.open(self.BASE_URL + '/@@add_content_type')
        browser.getControl('Title').value = u'Content Type Title'
        browser.getControl('Example Text').value = u'some Text'
        browser.getControl('save').click()
        assert "Item was added." in browser.contents


kotti_quiz/resources.py
from kotti.resources import Content
from sqlalchemy import Column
from sqlalchemy import ForeignKey
from sqlalchemy import Integer
from sqlalchemy import Unicode
from sqlalchemy import String

from kotti_quiz import _


class Quiz(Content):
    """Quiz Content type."""

    id = Column(Integer, ForeignKey('contents.id'), primary_key=True)

    type_info = Content.type_info.copy(
        name=u'Quiz',
        title=_(u'Quiz'),
        add_view=u'add_quiz',
        addable_to=[u'Document'],
        )


class Question(Content):
    """Question Content type."""

    id = Column(Integer, ForeignKey('contents.id'), primary_key=True)
    correct_answer = Column(Unicode(256))
    question_type = Column(String())

    # change the type info to your needs
    type_info = Content.type_info.copy(
        name=u'Question',
        title=_(u'Question'),
        add_view=u'add_question',
        addable_to=[u'Quiz'],
        )


class Answer(Content):
    """Answer Content type."""

    id = Column(Integer, ForeignKey('contents.id'), primary_key=True)

    type_info = Content.type_info.copy(
        name=u'Answer',
        title=_(u'Answer'),
        add_view=u'add_answer',
        addable_to=[u'Question'],
        )


kotti_quiz/views/content_type.py
# -*- coding: utf-8 -*-

"""
Created on 2014-09-22
:author: Andreas Kaiser (disko)
"""

import colander
from kotti.views.edit import AddFormView
from kotti.views.edit import ContentSchema
from kotti.views.edit import EditFormView
from pyramid.view import view_config
from pyramid.view import view_defaults


from kotti_quiz import _
from kotti_quiz.resources import Quiz
from kotti_quiz.resources import Question
from kotti_quiz.resources import Answer
from kotti_quiz.views import BaseView


class QuizSchema(ContentSchema):
    title = colander.SchemaNode(
        colander.String(),
        title = u'Quiz Bezeichnung',
        )


class QuestionSchema(ContentSchema):
    title = colander.SchemaNode(
        colander.String(),
        title = u'Question:',
        )
    correctanswer = colander.SchemaNode(
        colander.String(),
        title = u'Correct Answer:',
        )


class AnswerSchema(ContentSchema):
    title = colander.SchemaNode(
        colander.String(),
        title = u'Answering <Choice></Choice>:',
        )


@view_config(name=Quiz.type_info.add_view,
             permission='add',
             renderer='kotti:templates/edit/node.pt',)
class QuizAddForm(AddFormView):
    schema_factory = QuizSchema
    add = Quiz
    item_type = _(u"Quiz")


@view_config(name=Question.type_info.add_view,
             permission='add',
             renderer='kotti:templates/edit/node.pt',)
class QuestionAddForm(AddFormView):
    schema_factory = QuestionSchema
    add = Question
    item_type = _(u"Question")


@view_config(name=Answer.type_info.add_view,
             permission='add',
             renderer='kotti:templates/edit/node.pt',)
class AnswerAddForm(AddFormView):
    schema_factory = AnswerSchema
    add = Answer
    item_type = _(u"Answer")


@view_config(name='edit',
             context=Quiz, permission='edit',
             renderer='kotti:templates/edit/node.pt')
class QuizEditForm(EditFormView):
    schema_factory = QuizSchema


@view_config(name='edit',
             context=Question, permission='edit',
             renderer='kotti:templates/edit/node.pt')
class QuestionEditForm(EditFormView):
    schema_factory = QuestionSchema


@view_config(name='edit',
             context=Answer, permission='edit',
             renderer='kotti:templates/edit/node.pt')
class AnswerEditForm(EditFormView):
    schema_factory = AnswerSchema


@view_defaults(context=Quiz, permission="view")
class QuizView(BaseView):
    """View for Quiz Content Type"""

    @view_config(name="view",
                 request_method='GET',
                 renderer='kotti_quiz:templates/quizview.pt')
    def view_quiz(self):
        questions = self.context.children
        return {
            'questions': questions,
        }

    @view_config(name='view',
                 request_method='POST',
                 renderer='kotti_quiz:templates/checkview.pt')
    def check_answers(self):
        questions = self.context.children
        answers = self.request.POST
        sumtotal = 0
        sumcorrect = 0
        questioncorrect = {question.name: False for question in questions}
        for question in questions:
            sumtotal += 1
            #import pdb; pdb.set_trace()
            if question.name in answers:
                if question.correctanswer == answers[question.name]:
                    questioncorrect[question.name] = True
                    sumcorrect += 1

        return {
            'questions': questions,
            'questioncorrect': questioncorrect,
            'sumtotal': sumtotal,
            'sumcorrect': sumcorrect,
        }


@view_defaults(context=Question)
class QuestionView(BaseView):
    """View for Question Content Type"""

    @view_config(name="view", permission="view",
                 renderer='kotti_quiz:templates/questionview.pt')
    def view_question(self):
        answers = self.context.values()
        return {
            'answers': answers,
        }


kotti_quiz/views/quiz.py
# -*- coding: utf-8 -*-

"""
Created on 2014-09-22
:author: Andreas Kaiser (disko)
"""

import colander
from kotti.views.edit import AddFormView
from kotti.views.edit import ContentSchema
from kotti.views.edit import EditFormView
from pyramid.view import view_config
from pyramid.view import view_defaults
from deform.widget import RadioChoiceWidget

from kotti_quiz import _
from kotti_quiz.resources import Quiz
from kotti_quiz.resources import Question
from kotti_quiz.resources import Answer
from kotti_quiz.views import BaseView


class QuizSchema(ContentSchema):
    title = colander.SchemaNode(
        colander.String(),
        title=u'Quiz Bezeichnung',
        )


class QuestionSchema(ContentSchema):
    title = colander.SchemaNode(
        colander.String(),
        title=u'Question:',
        )
    correct_answer = colander.SchemaNode(
        colander.String(),
        title=u'Correct Answer:',
        missing=None,
        default=colander.null
        )
    question_type = colander.SchemaNode(
        colander.String(),
        validator=colander.OneOf(["radio", "checkbox", "text"]),
        widget=RadioChoiceWidget(values=[
            ["radio", "singlechoice"],
            ["checkbox", "multiplechoice"],
            ["text", "freetext"]])
        )


class AnswerSchema(ContentSchema):
    title = colander.SchemaNode(
        colander.String(),
        title=u'Answering Choice:',
        )


@view_config(name=Quiz.type_info.add_view,
             permission='add',
             renderer='kotti:templates/edit/node.pt',)
class QuizAddForm(AddFormView):
    schema_factory = QuizSchema
    add = Quiz
    item_type = _(u"Quiz")


@view_config(name=Question.type_info.add_view,
             permission='add',
             renderer='kotti:templates/edit/node.pt',)
class QuestionAddForm(AddFormView):
    schema_factory = QuestionSchema
    add = Question
    item_type = _(u"Question")


@view_config(name=Answer.type_info.add_view,
             permission='add',
             renderer='kotti_quiz:templates/answeradd.pt',)
class AnswerAddForm(AddFormView):
    schema_factory = AnswerSchema
    add = Answer
    item_type = _(u"Answer")

    @property
    def success_url(self):
        return self.request.resource_url(self.context)


@view_config(name='edit',
             context=Quiz, permission='edit',
             renderer='kotti:templates/edit/node.pt')
class QuizEditForm(EditFormView):
    schema_factory = QuizSchema


@view_config(name='edit',
             context=Question, permission='edit',
             renderer='kotti:templates/edit/node.pt')
class QuestionEditForm(EditFormView):
    schema_factory = QuestionSchema


@view_config(name='edit',
             context=Answer, permission='edit',
             renderer='kotti:templates/edit/node.pt')
class AnswerEditForm(EditFormView):
    schema_factory = AnswerSchema


@view_defaults(context=Quiz, permission="view")
class QuizView(BaseView):
    """View for Quiz Content Type"""

    @view_config(name="view",
                 request_method='GET',
                 renderer='kotti_quiz:templates/quizview.pt')
    def view_quiz(self):
        questions = self.context.children
        return {
            'questions': questions,
        }

    @view_config(name='view',
                 request_method='POST',
                 renderer='kotti_quiz:templates/checkview.pt')
    def check_answers(self):
        questions = self.context.children
        answers = self.request.POST
        sumtotal = 0
        sumcorrect = 0
        questioncorrect = {question.name: False for question in questions}
        for question in questions:
            sumtotal += 1
            #import pdb; pdb.set_trace()
            if question.name in answers:
                if question.correct_answer == answers[question.name]:
                    questioncorrect[question.name] = True
                    sumcorrect += 1

        return {
            'questions': questions,
            'questioncorrect': questioncorrect,
            'sumtotal': sumtotal,
            'sumcorrect': sumcorrect,
        }


@view_defaults(context=Question)
class QuestionView(BaseView):
    """View for Question Content Type"""

    @view_config(name="view", permission="view",
                 renderer='kotti_quiz:templates/questionview.pt')
    def view_question(self):
        answers = self.context.values()
        return {
            'answers': answers,
        }
